wasp/draw565.py

```diff
@@ -355,3 +355,47 @@ def line(n):
             chunks.append(end)
 
         return chunks
+
+    def line(self, x0=0, y0=0, x1=0, y1=0, color=0xffff):
+        """Draw a line between points (x0;y0) and (x1;y1) with color.
+
+        Example:
+
+        .. code-block:: python
+
+            draw = wasp.watch.drawable
+            draw.line(0, 120, 240, 240, 0xf800)
+
+        :param x0: X coordinate of the start of the line, defaults to 0
+        :param x1: Y coordinate of the end of the line, defaults to 0
+        :param y0: Y coordinate of the start of the line, defaults to 0
+        :param y1: Y coordinate of the end of the line, defaults to 0
+        :param color: Color to draw line in, in RGB565 format, defaults to 0xffff
+        """
+        def _pix(x, y, color):
+            self._display.set_window(x, y, 1, 1)
+            self._display.write_data(((color >> 8) & 0xFF, color & 0xFF))
+            # no need to set_window() again, just wastes time.
```

Comment:
This function is only called from one place so you'd probably be better off caching the internals of this method in local variables and then expanding the actual calls to set_window and write_data below:

~~~
set_window = self._display.set_window
write_data = self._display.write_data
px = bytes((color >> 8) & 0xff, color & 0xff))
~~~

This will remove the memory allocation (and several dictionary lookups) out of the inner loop.

Avoiding dictionary lookups doesn't make *that* much difference to benchmarks but avoiding allocations can. Talking of benchmarks can you add a simple line drawing benchmark into the self tests... doesn't need to be sophisticated just draw a bunch of lines in 15 degree increments and measure how long it takes.

Reply:
will add benchmarks later, implemented the other changes.

Reply:
Added benchmark and fixed bug with the vertical and horizontal lines optimization